packages/babel-parser/src/plugins/typescript/scope.js

```diff
@@ -0,0 +1,94 @@
+// @flow
+
+import ScopeHandler, { Scope } from "../../util/scope";
+import {
+  BIND_KIND_TYPE,
+  BIND_FLAGS_TS_ENUM,
+  BIND_FLAGS_TS_EXPORT_ONLY,
+  BIND_KIND_VALUE,
+  BIND_FLAGS_CLASS,
+  type ScopeFlags,
+  type BindingTypes,
+} from "../../util/scopeflags";
+import * as N from "../../types";
+
+class TypeScriptScope extends Scope {
+  types: string[] = [];
+
+  // enums (which are also in .types)
+  enums: string[] = [];
+
+  // classes (which are also in .lexical) and interface (which are also in .types)
+  classes: string[] = [];
+
+  // namespaces and bodyless-functions are too difficult to track,
+  // especially without type analysis.
+  // We need to track them anyway, to avoid "X is not defined" errors
+  // when exporting them.
+  exportOnlyBindings: string[] = [];
+}
+
+// See https://github.com/babel/babel/pull/9766#discussion_r268920730 for an
+// explanation of how typescript handles scope.
+
+export default class TypeScriptScopeHandler extends ScopeHandler<TypeScriptScope> {
+  createScope(flags: ScopeFlags): TypeScriptScope {
+    return new TypeScriptScope(flags);
+  }
+
+  declareName(name: string, bindingType: BindingTypes, pos: number) {
+    const scope = this.currentScope();
+    if (bindingType & BIND_FLAGS_TS_EXPORT_ONLY) {
+      this.maybeExportDefined(scope, name);
+      scope.exportOnlyBindings.push(name);
+      return;
+    }
+
+    super.declareName(...arguments);
+
+    if (bindingType & BIND_KIND_TYPE) {
+      if (!(bindingType & BIND_KIND_VALUE)) {
+        // "Value" bindings have already been registered by the superclass.
+        this.checkRedeclarationInScope(scope, name, bindingType, pos);
+        this.maybeExportDefined(scope, name);
+      }
+      scope.types.push(name);
+    }
+    if (bindingType & BIND_FLAGS_TS_ENUM) scope.enums.push(name);
+    if (bindingType & BIND_FLAGS_CLASS) scope.classes.push(name);
+  }
+
+  isRedeclaredInScope(
+    scope: TypeScriptScope,
+    name: string,
+    bindingType: BindingTypes,
+  ): boolean {
+    if (scope.enums.indexOf(name) > -1) {
+      // Enums can be merged with other enums
+      return !(bindingType & BIND_FLAGS_TS_ENUM);
```

Comment:
Enums can merge with namespaces too; I guess the error is avoided by not calling `super.declareName` for namespaces?

Reply:
Yeah exactly.